WIP - new crone, standalone jack client

catfact opened the pull request:
`norns/crone` contains a new jack client designed to replace some of the Crone supercollider components.

it has 4 inputs and 2 outputs. the idea is that the first pair of inputs are to be connected to system capture ports, 2nd pair is connected to supercollider playback, and outputs are connected to system playback.

`crone` will perform mixing and routing, host "built-in" effects and the `softcut` engine.

this is WIP:
- currently, there is only basic routing: ADC mix to monitor bus, bus to DAC, ext input to DAC.
- the interface is still OSC, which is overkill.
- none of the necessary changes to supercollider classes have been made.

will make some roadmap issues and update this PR accordingly.

but for now:
- the very first step is for  people to try building and running this project on the norns (requirements: `cmake`, `libboost-dev`, `libjack-dev`)

- i will be stitching fx and softcut into here shortly

- `norns/sc/core` needs to be updated; engines can send directly to jack client output ports.

- need to programmatically connect jack clients `supernova` -> `crone` (and disconnect `supernova`->system)

- as an interim measure, we could make an OSC bridge in `sclang` for routing/level commands; or, go straight to making changes in `matron/oracle`.

tehn commented on crone/src/JackClient.cpp:
```diff
@@ -0,0 +1,200 @@
+
+using namespace crone;
+
+class JackClient::Imp {
```
imp!!! 😸

catfact commented:
i was just gonna roll it into the main repo, already added the sources..

tehn commented:
forgive my CMake ignorance. i'm getting this error when running `cmake ..` from `crone/build/`

```
CMake Error at CMakeLists.txt:2 (project):
  project with VERSION must use LANGUAGES before language names.
```

catfact commented:
oh - hm, probably cmake version issue, i'll have a look. been using 3.13 on the laptop

catfact commented:
ok, this is getting somewhere usable and is worth giving a spin. new changes:
- everything uses `waf` now, no extra build steps
- changed `crone.sh` to explicitly launch supernova and the new `crone` client.
- changed `Crone` to use the remote supernova server, and connect it to system and `crone` client via `jack_connect`
- changed `CroneEffects`, now does no processing, forwards mix/fx param commands to `crone` client 
- changed `CroneAudioContext`, now doesn't have output busses/synth but uses the playback ports directly; forwards mix  commands to `crone` client.

so, everything basically works as before, but now mixing and FX are done entirely in a separate jack client. 

engines should work exactly as before; the `context` is still there and its `in_b` and `out_b` busses look identical, but now `out_b` just wraps the main outputs for the SC server.

CPU usage for the effects seems approximately the same or a little less (~20% for both?) but this setup seems to cure supernova's conniptions with `ParGroup`s feeding a heavy ugen. (in other words: now CPU usage drops as expected when switching from a heavy polysynth to a lighter engine.)

-----------

## caveats:

the monitor and input->aux routing is a little different (it could be changed back)
- now the aux fx chain takes input from supernova output (same as before) and from the monitor bus (instead of the ADC bus directly.)
- monitor mix is defined as 4 levels (L->L, L->R, &c) instead of 2 pans, but the control surface is unchanged.

## next steps / TODO:
- softcut. i'm undecided whether this is best as a separate jack client (could be in the same process) or actually baked into `crone`. i'm tempted to do the latter, since  it would be pretty cool to have it always accessible and able to mix engine output into its input. 
  - could have flags to enable/disable processing per voice.
  - could compute softcut on separate thread, at the expense of an extra buffer cycle of latency (same as for the multi-client arrangment if i understand correctly.)

  - tape. i am thinking of just using an off-the-shelf solution, in particular [ecasound](https://ecasound.seul.org/ecasound/Documentation/users_guide/html_uguide/users_guide.html#htoc7) -this is a lightweight but powerful record/playback utility that comes with an [API]((http://nosignal.fi/ecasound/Documentation/programmers_guide/ecasound_programmers_guide.html#application-development-using-the-ecasound-framework)) for integration with other programs. this would easily give us a lot of utiltiy (including multitrack / overdubbing, if desired) with little technical risk.

catfact commented:
> highly (highly) in favor of including softcut within crone.

kinda figured. already added it on this branch:
https://github.com/catfact/norns/tree/new-crone-softcut

which is completely untested (i must admit to feeling a certain amount of dread...)

i'd like to test that and this in parallel, in case there are fundamental issues here already. 

> re processing flags. i think a stop/start message would effectively work and remain intuitive?

i added a command to set processing flag per softcut voice. when disabled, both processing and mixing for that voice is bypassed. sure, can add start/stop command for all voices.

> i still think 4 mono softcut heads is best/reasonable. how do we envision the mix matrix?

ok. i dunno if you missed my comments on this, but i would be able to make a stereo version of a `SoftCutVoice` pretty easily. the approach i'd use would be making a variant head structure that is 'synced' to another head, and doesn't perform any loop or position logic (only read and write.) this would make the ratio of stereo processing load : mono proc load a lot better than 2:1, though still quite substantially worse than 1:1 ( b.c  r/w resampling are the most expensive operations.) maybe 1.3:1, ish?

anyways, currently the softcut mix structure is:
- 4 mono input buses, 4 mono output buses
- 4x2 array of ADC input levels
- 4x2 array of "ext" input levels (from supernova)
- 4x1 array of output levels
- 4x1 array of output pan positions
- 4x4 array of feedback levels (applied before output level+pan)
- stereo bus with mix of all softcut voices (after level+pan)

(replace 4 by any number of voices. actually j/k, it's 2 right now for testing purposes.)

> having gain control over input into supernova
> being able to direct alternative streams to supernova for processing, ie softcut output or tape from disk

i've no opinion on this really. happy to add loopback connections. it's of course pretty simple for SC to do its own input scaling or playback from disk.

> EQ is interesting to me. where in the chain would it go? post-compressor? 

i was thinking on the output/insert bus, pre-compressor, preserving potential for using compressor as a proper hard output limiter. but i dunno. we can experiment with this later.

> are you still considering the soft clipper anywhere?

there is a TSQ soft clipper inside each softcut subhead, applied after resampling and before buffer store. its purpose is mostly to guard against hard clipping from interpolation overshoot, less as a coloration effect. but we could expose its parameters. (there is a small but fun math problem involved in refactoring the coefficient calculation to give it a single gain parameter and a constant clipping point near unity.)

for saturation as coloration, i think there are maybe better transfer functions. TSQ is good at being totally linear below the knee; other functions have arguably nicer harmonic structures but are nonlinear all the way down to zero, more like a tube.

> [re:tape] lightweight it best? i'd keep it really simple.

sure ok, we can just roll our own then. but nothing is "really simple." at minimum we want to report current tape position, perform playback and recording. this is not a huge amount of work but requires some more thread management and care. i thought i'd mention the possibility of just using something more full-featured and never having to worry about adding our own features. NBD.

like,
> playback just needs the bare basics.

what are those?

> does jack provide level control at connection points? ie, master output and input? 

not that i'm aware of. mixing in crone is not a big deal at all. if it seems like the level smoothing is getting out of hand, we can arbitrarily make some mix levels static, or drop it down to smoothing at the block level.

that reminds me of another caveat.

## more caveats

### control rate

 currently, parameter changes are applied only once per audio block (though resulting values are smoothed at audio rate for e.g. softcut rate, pre/rec levels, mix point levels.) in other words there's no separate period analogous to SC or max kr / vector size. since block size is 128, this currently means about 3ms jitter on param changes. 

we could add a separate control processing rate. it's straightforward but fiddly, i'd like to get the basics working first. it would also be possible to support sample-accurate param changes, assuming incoming commands are timestamped, which is a little more work (though actually less fiddly in some ways.)

### polls

i need to add crone->matron plumbing. will probably stick with `Polls` class to complement `Commands`. commands use a SPSC lockfree queue to get into the audio thread; don't see a reason not to use a similar structure to get stuff out (most importantly, softcut phase and triggers, but also levels i guess.)

### soundfiles

obvs need to add a little library to deal with reading/writing sndfiles to/from RAM. maybe we could crib from some existing library, but it's maybe a good opportunity to do some things "right", like xfading when pasting new material in the middle of a buffer, &c.

tehn commented:
> i added a command to set processing flag per softcut voice. when disabled, both processing and mixing for that voice is bypassed. sure, can add start/stop command for all voices.

command for processing flag, yes. that sounds perfect. no need to add anything then.

> i would be able to make a stereo version of a SoftCutVoice pretty easily...

would this just be another flag? ie, dual stereo (2x2) or quad mono (4x1) mode?

> happy to add loopback connections. it's of course pretty simple for SC to do its own input scaling or playback from disk.

agreed that it's easy to have these things in sc (or whatever external application) but i think that grouping types of parameters makes sense (for input gain particularly). and also means not having to re-implement them in another application (ie, if someone wants to make pd-crone or whatever).

> but nothing is "really simple." at minimum we want to report current tape position, perform playback and recording. this is not a huge amount of work but requires some more thread management and care. i thought i'd mention the possibility of just using something more full-featured and never having to worry about adding our own features. NBD.

oh i'm totally ok with using something existing and full-featured! and yes you're correct that nothing is simple, didn't mean to come across as dismissive, ie

> > playback just needs the bare basics.
>
> what are those?

load, play, stop. could get crazy and add pause, resume, move to position. need to be able to get current position.

record is just specify file location, start, stop. report elapsed time.

> if it seems like the level smoothing is getting out of hand, we can arbitrarily make some mix levels static, or drop it down to smoothing at the block level.

agreed. some routings probably make more sense to be fixed, without need for smoothing. i'll draw some diagrams and try to identify these.

> control rate

i will defer to your judgement on this

> polls

👍

catfact commented:
> >   i would be able to make a stereo version of a SoftCutVoice pretty easily...

> would this just be another flag? ie, dual stereo (2x2) or quad mono (4x1) mode?

well, i really don't know. my main point is that stereo processing in itself could be cheap enough to always leave it on, and just change the input mix settings.

catfact commented:
added "built in" 2x softcut. it works. 

 _new_ caveats:

- CPU use is a bit worse than i'd hoped. with 2x softcut voices, reverb and compression, usage for `crone` is pushing 80% (single thread) and there are some xruns. i'll of course be digging into this more. basically every single option that uses more CPU is "on" right now. (i _still_ haven't buckled down and gotten to grips with using `prof` on the norns, and that is basically a prerequisite for any kind of intelligent optimization.)

- ... oh, except the feedback mix. that's disabled for the moment.

- also note that there is no glue from matron -> softcut. the OSC interface for softcut params/routing is in there, but right now it can only be controlled remotely (see `crone/osc-tests/softcut.scd`)

still, it is quite exciting to boot up and run `awake` into the softcut.

catfact commented:
added multithreaded DSP; now softcut processing is computed on a worker.

this basically works. however (see commit message) the timing isn't quite right yet, and there is a good chance of glitches happening that are basically buffer underruns but aren't reported as such by jack.

bumping jack buffer size to 256 ameliorates this, but of course it is not ideal. will continue working on it.

here's the current state of things. i'm able to run 2x softcut voices on top of `loom` with 8 voices, reverb and compression enabled. this is with bufsize=256. [*]


```
top - 03:03:47 up 32 min,  5 users,  load average: 1.08, 0.94, 0.74
Threads: 167 total,   3 running, 122 sleeping,   0 stopped,   0 zombie
%Cpu0  : 24.5 us,  2.3 sy,  0.0 ni, 72.9 id,  0.0 wa,  0.0 hi,  0.3 si,  0.0 st
%Cpu1  : 40.8 us,  1.9 sy,  0.0 ni, 57.2 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu2  : 49.5 us,  1.0 sy,  0.0 ni, 49.5 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu3  : 50.7 us,  2.6 sy,  0.0 ni, 46.7 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem :  1002896 total,   524464 free,   178948 used,   299484 buff/cache
KiB Swap:        0 total,        0 free,        0 used.   669260 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S %CPU %MEM     TIME+ COMMAND                                                                                                                                          
 9686 we        20   0  179360 156956  83880 R 43.8 15.7   3:45.40 crone                                                                                                                                            
 9689 we       -91   0  179360 156956  83880 R 35.5 15.7   2:00.70 crone                                                                                                                                            
 9717 we        20   0  132428  10588   7720 S 19.1  1.1   0:15.05 matron                                                                                                                                           
 9681 we       -91   0  182812 116512  95604 S 16.8 11.6   1:03.26 DSP Thread 0                                                                                                                                     
 9682 we       -91   0  182812 116512  95604 S 14.8 11.6   0:11.91 DSP Thread 1                                                                                                                                     
 9683 we       -91   0  182812 116512  95604 S 14.8 11.6   0:11.80 DSP Thread 2                                                                                                                                     
 9684 we       -91   0  182812 116512  95604 S 14.8 11.6   0:11.66 DSP Thread 3   
```

couple things to notice:
- two threads marked `crone`; one is doing softcut process, the other is doing everything else (mix, fx)
-  `matron` activity has become significant with the amount of stuff going on in `loom` at high bpm and high voice count.  

[*] to change bufsize, something like:
- edit `/etc/systemd/system/norns-jack.service`
- `systemctl daemon-reload`
- `systemctl restart norns-jack.service`

catfact commented:
PS - i'm weighing the option of just going back to `softcut` as a separate jack client. it could still get input from supernova. 
headache of managing DSP threads internally goes away. 
new headache appears in the form of a separate agent for commands/polls. 
opinions welcome.

catfact commented:
ok, implemented the multi-client, single-process arrangement. 

see comments on commit  f022e1b. 

this is the arrangement where all signals from ADC/to DAC are routed through crone mixer.

this creates feedback loops in jack (crone -> ext -> crone, crone -> softcut -> crone.) i'm not sure of the exact ramifications of doing this (re: latency, block callback scheduling.) it's simple to switch to a more independent arrangement if needed (where ext/softcut directly connect ADC); the current version has more mix points, but has the advantage of having a global ADC level controlled by a single mix point.

catfact commented:
i could use help testing this on norns. also the scsynth -> crone connection isn't correct right now - it happens in sclang after server boot, but needs to connect crone output 4,5 -> scsynth, scsynth output -> crone input 4,5.

here's the jack connection graph without scsynth:

![screenshot from 2018-11-29 21-02-14](https://user-images.githubusercontent.com/94855/49263843-14cb6080-f41a-11e8-8d89-24d4cd5b6dcf.png)

tehn commented:
holy moly this is exciting.

i'll begin looking through the code and will test on hardware.

initial quick questions:

- are all of the mix connections managed by OSC forwarding via sclang?
- how is softcut controlled currently? similarly with OSC forwarding?

tehn commented:
oh or they're just controlled via osc directly ok

catfact commented:
the `crone` process creates two jack clients (mixing/fx, softcut) and a single OSC interface (port number, server thread) that controls both of them.

> are all of the mix connections managed by OSC forwarding via sclang?

the ones that existed in AudioContext  are, the new ones haven't been added yet. see here
https://github.com/catfact/norns/blob/new-crone/sc/core/CroneAudioContext.sc#L114

and here
https://github.com/catfact/norns/blob/new-crone/sc/core/CroneEffects.sc#L23

> how is softcut controlled currently? similarly with OSC forwarding?

currently it can only be controlled remotely. i'll add something presently. i guess for minimum friction with scripts it should be  some kind of Engine wrapper. i have a vague idea that will work and will see how it goes.

obvs i also need to add polls and soundfile management.

catfact commented:
ach, apologies, i haven't updated wscript yet, 1sec

catfact commented:
@tehn i've  gone ahead and merged this set of changes even though its incomplete:
- it feels backwards to be building an OSC glue wrapper engine in sclang (esp. since i coudn;t get it working right away for some reason)

- so, started just adding glue directly to oracle. so far just done most of the levels/routing commands. not fx or softcut parameters. but i put the needed data in the comments, it is just a few editor macros away, too late at night now to continue.

- hm also this highlights a few issues and decisions made in the jack client that are kinda 'weird' (e.g. decision to explicitly reference fx algo and param names in the OSC paths) and also some ways in which new mix structure differs from old, for better or worse (e.g. ADC channels now considered as stereo pair; aux send comes after monitor mix instead of before)

but anyways i wonder if you think it's too tricky to have to refactor the lua scripts using softcut right now. (mlr, cranes,... anything else?) if we can do that i think it would be good to just go matron->crone directly. not least b/c it is then easy for a lua script to load a scsynth engine and also use softcut.

tehn commented:
softcut is priority. mlr can be completely redesigned/redone, and i'll help dan derks fix crane (or do it for him) regarding any changes... so please do not be concerned with lua script compatibility.

agreed that softcut should be persistent and not involved in engine-space, almost like an extension to the `audio` parameters. i'm pretty sure this is what you're thinking already.

do you have an idea how to implement the softcut polls since they'll be always available? just have callback similar to `norns.vu`? vs. using the dynamic poll system etc.

catfact commented:
oh but you're talking about the lua interface huh. will probably end up still providing a way to enable/disable phase and loop polls (whatever the IPC interface happens to be,) because there will be some measurable performance impact from computing these on the backend.

IOW, they will look like the pitch polls as far as lua is concerned.

catfact commented:
BTW, i did get `perf` set up and ran some samples. nothing too unexpected; 32% of `crone` work is spent in `poke()`. 9% in `peek()`.  feedback mix is surprisingly huge at 12%. i suspect denormals and will try crafting an appropriate gremlin-zapper for ARM.

catfact commented:
let me know if there's anything i can explain or document better right now.

as it stands, probably the most significant thing is just the OSC interface. all methods that the `crone` process accepts are printed on launch, and collected in `crone/osc-methods.txt`. (which is manually updated with `crone` output, so it may be stale.)

catfact commented:
currently i'm adding some basic buffer/soundfile management for softcut.

tehn commented on matron/src/oracle.h:
```diff
@@ -141,5 +141,16 @@ extern void o_set_insert_fx_off();
 extern void o_set_insert_fx_param(const char* name, float value);
 
+//--- softcut parameter controls
+//--- TODO: no lua glue for these yet
```
it does seem like a good division. i'll handle `weaver` implementation then.

agreed that a merge to `monome/new-crone` would be nice as i tend to work on many different norns hardwares with different git setups :/
